src/scaffoldmaker/meshtypes/meshtype_3d_bladder1.py

```diff
@@ -880,13 +573,676 @@ def defineFaceAnnotations(cls, region, options, annotationGroups):
+
+        is_neck_lumen_dorsal = fm.createFieldAnd(is_neck_lumen, is_dorsal_bladder)
+        lumenOfNeck_dorsal = \
+            findOrCreateAnnotationGroupForTerm(annotationGroups, region,
+                                               get_bladder_term("dorsal part of lumen of neck of urinary bladder"))
+        lumenOfNeck_dorsal.getMeshGroup(mesh2d).addElementsConditional(is_neck_lumen_dorsal)
+
+        is_neck_lumen_ventral = fm.createFieldAnd(is_neck_lumen, is_ventral_bladder)
+        lumenOfNeck_ventral = \
+            findOrCreateAnnotationGroupForTerm(annotationGroups, region,
+                                               get_bladder_term("ventral part of lumen of neck of urinary bladder"))
+        lumenOfNeck_ventral.getMeshGroup(mesh2d).addElementsConditional(is_neck_lumen_ventral)
+
+def findDerivativeBetweenPoints(v1, v2):
+    """
+    Find vector difference between two points and rescale vector difference using cubic hermite arclength
+    between the points to derive the derivative between the points.
+    :param v1: start vector
+    :param v2: end vector
+    :return: derivative of between v1 and v2
+    """
+    d = [v2[c] - v1[c] for c in range(3)]
+    arcLengthAround = interp.computeCubicHermiteArcLength(v1, d, v2, d, True)
+    d = [c * arcLengthAround for c in vector.normalise(d)]
+
+    return d
+
+def findNodesAlongBladderDome(sx_dome_group, elementsCountAround, elementsCountAlongDome):
+
+    # Find apex nodes d2
+    d2Apex = []
+    d2 = sx_dome_group[2][0]
+    for n1 in range(elementsCountAround):
+        rotAngle = n1 * 2.0 * math.pi / elementsCountAround
+        rotAxis = vector.normalise(vector.crossproduct3(vector.normalise(sx_dome_group[2][0]), vector.normalise(sx_dome_group[4][0])))
+        rotFrame = matrix.getRotationMatrixFromAxisAngle(rotAxis, rotAngle)
+        d2Rot = [rotFrame[j][0] * d2[0] + rotFrame[j][1] * d2[1] + rotFrame[j][2] * d2[2] for j in range(3)]
+        d2Apex.append(d2Rot)
+
+    # Create ellipses along dome around the central path
+    xEllipses_dome = []
+    d1Ellipses_dome = []
+    for n in range(1, len(sx_dome_group[0])):
+        px, pd1 = createEllipsePoints(sx_dome_group[0][n], 2 * math.pi, sx_dome_group[2][n], sx_dome_group[4][n], elementsCountAround,
+                                      startRadians=0.0)
+        xEllipses_dome.append(px)
+        d1Ellipses_dome.append(pd1)
+
+    # Find d2
+    d2Raw = []
+    for n1 in range(elementsCountAround):
+        xAlong = []
+        d2Along = []
+        for n2 in range(len(xEllipses_dome) - 1):
+            v1 = xEllipses_dome[n2][n1]
+            v2 = xEllipses_dome[n2 + 1][n1]
+            d2 = findDerivativeBetweenPoints(v1, v2)
+            xAlong.append(v1)
+            d2Along.append(d2)
+        xAlong.append(xEllipses_dome[-1][n1])
+        d2Along.append(d2)
+        d2Smoothed = interp.smoothCubicHermiteDerivativesLine(xAlong, d2Along)
+        d2Raw.append(d2Smoothed)
```

Comment:
The above code should get an exact measure of the d2 direction at start and end. I suggest getting d2, d12 and d3, d13 at the body/neck junction, which should be on a node in the central path. Use d12 and d13 to linearly scale d2 and d3 for a small distance "delta" in the direction of d1 & add another ellipse there; you can use the difference in coordinates between this ellipse and the actual last ellipse just like your above code.
You also need to put the exact direction in at the apex.
Use fixStartDirection=True, fixEndDirection=True in the smooth call.
(If you turn off lines and show solid surfaces you can see a rather sudden lighting change at the Dome/Neck junction & the above should help to fix it; the Neck code also needs to play along.)